File: 1756-minimum-deletions-to-make-string-balanced/minimum-deletions-to-make-string-balanced.cpp
class Solution {
public:
    int minimumDeletions(string s) {
        int sz = s.length(), ans = INT_MAX;
        vector<int> a(sz, 0), b(sz, 0);
        int lastValue = 0;

        for(int i=0; i<sz; i++){
            if(s[i] == 'b'){
                a[i] = lastValue++;
            }
            else a[i] = lastValue;
        }
        if(lastValue == sz || lastValue == 0) return 0;
        lastValue = 0;

        for(int i=sz-1; i>=0; i--){
            if(s[i] == 'a'){
                b[i] = lastValue++;
            }
            else b[i] = lastValue;
        }

        for(int i=0; i<sz; i++){
            // cout << a[i] << ":" << b[i] << endl;
            ans = min(ans, a[i]+b[i]);
        }

        return ans;
    }
};

File: 1756-minimum-deletions-to-make-string-balanced/minimum-deletions-to-make-string-balanced.java
class Solution {
    public int minimumDeletions(String s) {
        int sz = s.length(), a = 0, b = 0, ans = Integer.MAX_VALUE;

        for(int i=0; i<sz; i++){
            if(s.charAt(i) == 'a') a++; 
            // else a++; 
        }

        for(int i=0; i<sz; i++){
            if(s.charAt(i) == 'a') a--; 
            // else b++;
            // System.out.println(i + " " + a + ":" + b);
            ans = Math.min(ans, a + b);
            if(s.charAt(i) == 'b') b++; 
        }

        return ans;
    }
}